io_scene_godot/converters/animation.py

```diff
@@ -0,0 +1,200 @@
+        self.data = collections.OrderedDict()
+        self.id = track_id
+        self.data['type'] = '"{}"'.format(track_type)
+        self.data['path'] = node_path
+        self.data['interp'] = 1
+        self.data['keys'] = KeyFrameList(track_type)
+
+    def append_keyframe(self, keyframe):
+        """Append a keyframe to the track if the last
+        keyframe in the track is different"""
+        # if the keyframe is identical to the last keyframe
+        # it does not need to be exported
+        if (not self.data['keys'] or
+                not KeyFrame.identical(self.data['keys'][-1], keyframe)):
+            self.data['keys'].append(keyframe)
+
+
+def attach_tracks_to_anim_data(anim_data, track_list, length):
+    """Add track data into a animtion resource"""
+    anim_data['loop'] = False
+    anim_data['step'] = 0.1
+    anim_data['length'] = length
+    track_list = sorted(track_list, key=lambda x: x.id)
+    for track in track_list:
+        for key, value in track.data.items():
+            anim_data['tracks/{}/{}'.format(track.id, key)] = value
+
+
+def extract_anim_objects(gd_node, anim_objects):
+    """Extract all the node.anim_object in the descendants of
+    a Godot node(includeing itself) into a {path:object} map"""
+    if gd_node.anim_object is not None:
+        anim_objects[gd_node.get_path()] = gd_node.anim_object
+        gd_node.anim_object = None
+    for child in gd_node.children:
+        extract_anim_objects(child, anim_objects)
+
+
+def export_animation_data(escn_file, name, anim_objects, anim_root_path):
+    """Export animation as an internal resource"""
+    animation_resource = InternalResource('Animation')
+
+    # initilize track and frame range
+    frame_start = float('inf')
+    frame_end = float('-inf')
+    tracks = list()
+    for path, obj in anim_objects.items():
+        tracks.append(Track(
+            len(tracks),
+            NodePath(anim_root_path, path),
+            'transform'
+        ))
+
+        if obj.type.lower() == 'armature' and obj.pose is not None:
+            for pbone in obj.pose.bones:
+                tracks.append(Track(
+                    len(tracks),
+                    NodePath(anim_root_path, path, pbone.name),
+                    'transform'
+                ))
+
+        frame_range = obj.animation_data.action.frame_range
+        if frame_range[0] < frame_start:
+            frame_start = frame_range[0]
+        if frame_range[1] > frame_end:
+            frame_end = frame_range[1]
+
+    original_frame = bpy.context.scene.frame_current
+
+    frame_len = 1 / bpy.context.scene.render.fps
+    frame_start = int(frame_start)
+    frame_end = int(frame_end)
+
+    # export the animation is done by baking frame because
+    # blender keyframe transform is relative to world space
+    # while godot transform is relative to parent space
+    for frame in range(frame_start, frame_end+1):
+        bpy.context.scene.frame_set(frame)
```

Comment:
Stepping through the frames of the scene has several limitations, as it forces all of the animations to be active in the scene to be exported. In blender it is relatively standard to use an action library (eg to differentiate between a walk cycle and a run cycle). This concept is present in Godot as well (multiple animations in the same animationplayer) so even if this isn't supported now, it would be good to ensure the exporter can easily support it in the future.

As a result I would rather have some sort of function operating on the F-curves - even though it would have to evaluate them every frame.

Reply:
If we deal with F-curves, it means we need to convert the F-curve keyframe value from world space to parent space, the conversion would guarantee the keyframe perfectly exported, however, as the value of F-curve changed, the interpolation would be changed, thus the whole animation (the non-keyframe frames) would be different.

If I do not mistake it, the approach to create something like an animation library in blender is to use the nla_stack, and nla_stack can be accessed through object->animation_data->[nla_track]->[nla_strip]->action, it can be easily exported